### PR TITLE
Add alternate 3D display modes

### PR DESCRIPTION
**Pull request by headassbtw**
Adds different methods of displaying stereo 3D gameplay:
* Side-by-Side
* Over-Under
* Anaglyph

This Works on My Machine:tm: with the Vulkan renderer, OpenGL crashes on 3D no matter what, and fixing that is outside the scope of this PR, however I did put in the necessary changes for it to work on the OpenGL backend once 3D support for it comes back up.

### CODE REVIEW

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <QScreen>
 #include <QUrl>
 
+#include "Emu/system_config_types.h"
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:15]: remove this line. unnecessary and wrong placement anyway

File: rpcs3/Emu/system_config_types.h [hunk 2]
```diff
@@ -312,3 +312,10 @@ enum class output_scaling_mode
 	bilinear,
 	fsr
 };
+
+enum class stereo_render_mode_options
+{
+	anaglyph,
+	side_by_side,
+	over_under
+};
```
Comment (Megamouse) [inline rpcs3/Emu/system_config_types.h:321]: missing newline at end of file

File: rpcs3/Emu/system_config.h [hunk 3]
```diff
@@ -161,6 +161,7 @@ struct cfg_root : cfg::node
 		cfg::_bool multithreaded_rsx{ this, "Multithreaded RSX", false };
 		cfg::_bool relaxed_zcull_sync{ this, "Relaxed ZCULL Sync", false };
 		cfg::_bool enable_3d{ this, "Enable 3D", false };
+		cfg::_enum<stereo_render_mode_options> stereo_render_mode{ this, "3D Display Mode", stereo_render_mode_options::side_by_side };
```
Comment (Megamouse) [inline rpcs3/Emu/system_config.h:164]: I don't understand why you kept the bool...
Just add a first enum value that represents the 2D default

File: rpcs3/Emu/RSX/VK/VKPresent.cpp [hunk 4]
```diff
@@ -653,8 +654,9 @@ void VKGSRender::flip(const rsx::display_flip_info_t& info)
 
 			vk::get_overlay_pass<vk::video_out_calibration_pass>()->run(
 				*m_current_command_buffer, areau(aspect_ratio), direct_fbo, calibration_src,
-				avconfig.gamma, !use_full_rgb_range_output, avconfig._3d, single_target_pass);
+				avconfig.gamma, !use_full_rgb_range_output, avconfig._3d, avconfig.stereo_mode, single_target_pass);
 
+
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKPresent.cpp:659]: remove this line

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 5]
```diff
@@ -471,26 +471,59 @@ namespace gl
...
 			"uniform int stereo;\n"
 			"uniform int stereo_image_count;\n"
+			"uniform int stereo_display_method;\n"
 			"\n"
 			"vec4 read_source()\n"
 			"{\n"
 			"	if (stereo == 0) return texture(fs0, tc0);\n"
 			"\n"
-			"	vec4 left, right;\n"
+			"	vec4 left, right, final;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:480]: final is a keyword in c++. Please use something else

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 6]
```diff
@@ -471,26 +471,59 @@ namespace gl
...
 			"	if (stereo_image_count == 2)\n"
 			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
+			"   vec2 coord_left = tc0 * vec2(1.f, 0.4898f);\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:483]: The formatting in your added code is completely off.
The indentations are all over the place.
You're missing spaces in "if(" and after "//" in comments

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 7]
```diff
@@ -899,27 +899,60 @@ namespace vk
...
 			"{\n"
 			"	if (stereo == 0) return texture(fs0, tc0);\n"
 			"\n"
-			"	vec4 left, right;\n"
+			"	vec4 left, right, final;\n"
 			"	if (stereo_image_count == 2)\n"
 			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
+			"   vec2 coord_left = tc0 * vec2(1.f, 0.4898f);\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:912]: The formatting in your added code is completely off.
The indentations are all over the place.
You're missing spaces in "if(" and after "//" in comments

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 8]
```diff
@@ -899,27 +899,60 @@ namespace vk
...
 			"	int stereo_image_count;\n"
+			"	int stereo_display_method;\n"
 			"};\n"
 			"\n"
 			"vec4 read_source()\n"
 			"{\n"
 			"	if (stereo == 0) return texture(fs0, tc0);\n"
 			"\n"
-			"	vec4 left, right;\n"
+			"	vec4 left, right, final;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:909]: final is a keyword in c++. Please use something else

File: rpcs3/rpcs3qt/emu_settings.cpp [hunk 9]
```diff
@@ -1267,6 +1267,14 @@ QString emu_settings::GetLocalizedSetting(const QString& original, emu_settings_
 		case vk_exclusive_fs_mode::enable: return tr("Prefer exclusive fullscreen", "Exclusive Fullscreen Mode");
 		}
 		break;
+	case emu_settings_type::StereoRenderMode:
+		switch (static_cast<stereo_render_mode_options>(index))
+		{
+		case stereo_render_mode_options::anaglyph: return tr("Anaglyph", "3D");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/emu_settings.cpp:1273]:
```suggestion
		case stereo_render_mode_options::anaglyph: return tr("Anaglyph", "3D Display Mode");
```

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 10]
```diff
@@ -629,6 +630,18 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 	m_emu_settings->EnhanceComboBox(ui->outputScalingMode, emu_settings_type::OutputScalingMode);
 	SubscribeTooltip(ui->outputScalingMode, tooltips.settings.output_scaling_mode);
 
+	//3D
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:633]:
```suggestion
	// 3D
```

**Review by kd-11 — CHANGES_REQUESTED** [review]
Address review comments.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 1 above)
Comment (headassbtw) [reply rpcs3/rpcs3qt/settings_dialog.cpp:15]: sorry, my IDE did that

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp (hunk 6 above)
Comment (headassbtw) [reply rpcs3/Emu/RSX/GL/GLOverlays.cpp:483]: Indentation is weird, looks fine in vscode, however i think that's automatically accounting for the indentation already present, which i am struggling to figure out what it is.

File: rpcs3/Emu/RSX/VK/VKPresent.cpp [hunk 11]
```diff
@@ -460,9 +460,10 @@ void VKGSRender::flip(const rsx::display_flip_info_t& info)
 		if (!buffer_pitch)
 			buffer_pitch = buffer_width * avconfig.get_bpp();
 
-		const u32 video_frame_height = (!avconfig._3d? avconfig.resolution_y : (avconfig.resolution_y - 30) / 2);
-		buffer_width = std::min(buffer_width, avconfig.resolution_x);
-		buffer_height = std::min(buffer_height, video_frame_height);
+		//commented out because I don't see a point in doing this, however others may
```
[the remark this answers is not in the document]
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp:463]: By "squishing the framebuffer" i mean this:
Screenshot_20230327_081031
It's not set to anaglyph in that screenshot, but it displays like that anyway (i have a hunch it's bypassing something in the shader but i'll look into that after i finish addressing the rest of the review comments)

**Review by AAAAAXfda — COMMENTED** [review]
Hi there! I'll be reviewing your commits and then I'll make a decision. Thanks for your contribution to RPCS3!

**Review by AAAAAXfda — APPROVED** [review]
Code looks clean, everything else looks good so far. I'll be merging this now. Have a good day!

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 12]
```diff
@@ -887,50 +887,81 @@ namespace vk
...
+      "	int stereo_image_count;\n"
+      "	int stereo_display_method;\n"
+      "};\n"
+      "\n"
+      "vec4 read_source()\n"
+      "{\n"
+      "	if (stereo == 0) return texture(fs0, tc0);\n"
+      "\n"
+      "	vec4 left, right;\n"
+      "	if (stereo_image_count == 2)\n"
```
Comment (kd-11) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:910]: This whole section is wrong.
You can get 1 image or 2 with 3D. This is an internal detail.
PS3 renders over-under, but some games will write both eyes in one pass (so we have one texture cached) and some will have 2 passes (so we have 2 textures)
The read source method should not have been altered at all, the only thing you should change is how you write the outputs. Generate the inputs at the vertex stage, not the fragment stage.
i.e If you want SBS/Over-under, you should emit 2 draw calls, one for each eye.
The only reason this seems to have even worked seems to have been by chance, you were testing a game that emits 2 draw calls and got 2 separate images.

Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKOverlays.cpp:910]: I suggest to use multi-draw (instanced) and use shader_draw_parameters to pick which coordinates to emit.

To summarize:
- This fragment output shader should really have no logic to handle any of these, you can send those to the VS and place them correctly.
- When doing SBS/Over-Under you no longer have stereo output, so you need a new variable to only read source but composite for one eye only.
- VS should just have a set of coordinates as constants and just emit the relevant ones for the current mode and eye.

Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKOverlays.cpp:910]: ah, could you name a game that does the other method? i was testing with killzone 3.

also i'm not quite sure how i'd do that at the vertex stage, i could probably figure it out but having slightly more info when i'm able to work on it again wouldn't hurt.

File: rpcs3/Emu/RSX/VK/VKPresent.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: On looking at the shader, it is indeed completely wrong which is why this happens.

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp (hunk 12 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKOverlays.cpp:910]: Try motorstorm and pacific rift, it's been a long time, I cannot remember which game used which mode.

If doing it with verts is too hard, you can still do it with the fragment shader, though the logic exhibited in this PR is incorrect. The easiest way would be to use a compute shader with image_load_store instead of draw calls. The logic that cares about the stereo mode should be after the read_source() call where you already have a left and right color. You would then use a 2D matrix to place the fragment in its correct eye coordinate. Luckily, as we don't do warping or rotation, your matrix is a simple scale and translate matrix that should take fs0 in and give you a left and right coord.

To do it with a draw call pass is a reverse operation. Instead of reading fs0, you would do an inverse projection with the 2D matrix to get the original eye id (left/right) and the coordinate within the eye. You would then call read_source, select the color you want (left/right) and write to fragment out as usual.

File: rpcs3/Emu/RSX/VK/VKPresent.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: The question then becomes - why? I mean, obviously if the code exists there must be a reason.

[the remark this answers is not in the document]
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: probably to not stretch the framebuffer unless the user allows it with the option that expands to the window, but as far as i can tell that's different, so i'm not sure. the original 3D PR shows it filling the window, so something must have changed along the way.

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: I mean, the buffer_width should be (PS3_VERTICAL_STACKED_RES - 30) / 2 as the code does, so as far as I can see this code is doing its job. Is master also stretched in KZ3? It shouldn't be.

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp (hunk 12 above)
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKOverlays.cpp:910]: i'm not calculating it, it doesn't matter as it's not VR.

the VR viewing app i used displayed it fine, as a floating screen in front of me, with one half of the window in each eye respectively

Screenshot_20230326_022836

File: rpcs3/Emu/RSX/VK/VKPresent.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: yep, it is stretched on master:

Screenshot_20230327_110009

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: In that case leave it stretched, I'll have to figure out the real cause of that.

[the remark this answers is not in the document]
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: sorry, not quite sure what you meant by leaving it stretched:

squished to the framebuffer size, like on master, or stretched to the window like my original commit?

File: rpcs3/Emu/system_config.h (hunk 3 above)
Comment (headassbtw) [reply rpcs3/Emu/system_config.h:164]: i kept the bool because from what i saw, enabling 3D was a bit more than that bool, and i didn't want to step on any toes by removing it

File: rpcs3/Emu/RSX/VK/VKPresent.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: hold on though, except for anaglyph and maybe SBS, wouldn't making the framebuffer the half-height lose detail? what's the point in halving the size if we're gonna put twice as much info in it with over-under?

File: rpcs3/Emu/system_config.h (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/system_config.h:164]: It is only confusing to have more than one variable for the same thing. Either it's 2D or one of the 3D values.
Having an extra bool has no benefit and only adds more elements in all related code

File: rpcs3/Emu/RSX/VK/VKPresent.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKPresent.cpp]: That's not what that variable is for. buffer_width/height is an input specifier, not output. The game calibrates it as something like 1280x1470 for 720p 3D and we need to calculate backwards the size for each eye.

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 13]
```diff
@@ -469,28 +469,57 @@ namespace gl
...
+			"	if (stereo_display_mode == 0) return texture(fs0, tc0);\n"
 			"\n"
 			"	vec4 left, right;\n"
 			"	if (stereo_image_count == 2)\n"
 			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
+			"		vec2 coord_left = tc0 * vec2(1.f, 0.4898f);\n"
+			"		vec2 coord_right = coord_left + vec2(0.f, 0.510204f);\n"
+			"		if(stereo_display_mode == 1)\n" // anaglyph
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:484]: You can just format the enum values to a string representation in these comparisons.
There's no need to hardcode 1,2,3 etc.

**Comment by AARP41298** [comment]
Hi, i just test it with GoW Origins Collection
Master with anaglyph:
master now
This pull with anaglyph:
pull now
This pull with OU:
3d OU rpcs3
This pull with SBS:
3d SBS2 rpcs3

**Comment by kd-11** [comment]
> must be that alternate 3D mode that the games i tested on don't have,
> 
> i'll make a note of it, if the uncharted 3 dump i left FTPing doesn't work

Descope those changes from this PR. They will not be accepted lumped in with the other work.

**Comment by headassbtw** [comment]
i'm getting confused on what you want me to do at this point

**Comment by kd-11** [comment]
> i'm getting confused on what you want me to do at this point

I've been pretty clear from the start - the changes you made around the buffer_width/buffer_height calculations are incorrect. Remove them.

File: rpcs3/Emu/RSX/GL/GLPresent.cpp [hunk 14]
```diff
@@ -149,7 +149,8 @@ void GLGSRender::flip(const rsx::display_flip_info_t& info)
 		if (!buffer_pitch)
 			buffer_pitch = buffer_width * avconfig.get_bpp();
 
-		const u32 video_frame_height = (!avconfig._3d ? avconfig.resolution_y : (avconfig.resolution_y - 30) / 2);
+		//gonna come back to this once i settle on how to handle this with modes, it makes sense for anaglyph but not SBS or over-under
+		const u32 video_frame_height = (avconfig.stereo_mode == stereo_render_mode_options::anaglyph? (avconfig.resolution_y - 30) / 2 : avconfig.resolution_y);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLPresent.cpp:153]: I think he means this
```suggestion
		const u32 video_frame_height = (avconfig.stereo_mode == stereo_render_mode_options::disabled ? avconfig.resolution_y : (avconfig.resolution_y - 30) / 2);
```

Comment (headassbtw) [reply rpcs3/Emu/RSX/GL/GLPresent.cpp:153]: yeah i knew about that, just wasn't sure about the general other logic, maybe it'll become more acceptable once i get the modes working with the proper buffer size.

thanks for the clarification, i'll work on that, although it'll be a bit because i'm not home and VNC over a slow connection isn't ideal.

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 15]
```diff
@@ -887,50 +887,80 @@ namespace vk
...
-			"void main()\n"
-			"{\n"
-			"	vec4 color = read_source();\n"
-			"	color.rgb = pow(color.rgb, vec3(gamma));\n"
-			"	if (limit_range > 0)\n"
-			"		ocol = ((color * 220.) + 16.) / 255.;\n"
-			"	else\n"
-			"		ocol = color;\n"
-			"}\n";
+      "#version 420\n\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:890]: The indentation in this whole section is altered and adds a lot of unnecessary changes to this PR.

Comment (headassbtw) [reply rpcs3/Emu/RSX/VK/VKOverlays.cpp:890]: indentation issues are mostly due to vscode having a fun time, i'll be sure to fix them though.

**Comment by DxsSucuk** [comment]
👀

**Comment by headassbtw** [comment]
Was there anything else i needed to do? I changed what was requested and tweaked the shader stuff to be as non-jank as possible without changing draw calls (because frankly you guys know more than me about those.)

Although I left in an `else` that could be omitted or reworked, I'd like to get feedback on that.

File: rpcs3/Emu/RSX/GL/GLPresent.cpp [hunk 16]
```diff
@@ -269,7 +269,7 @@ void GLGSRender::flip(const rsx::display_flip_info_t& info)
 		// TODO: Implement FSR for OpenGL and remove this fallback to bilinear
 		const gl::filter filter = g_cfg.video.output_scaling == output_scaling_mode::nearest ? gl::filter::nearest : gl::filter::linear;
 
-		if (use_full_rgb_range_output && rsx::fcmp(avconfig.gamma, 1.f) && !avconfig._3d)
+		if (use_full_rgb_range_output && rsx::fcmp(avconfig.gamma, 1.f) && avconfig.stereo_mode != stereo_render_mode_options::disabled)
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLPresent.cpp:272]: I think this is a typo
```suggestion
		if (use_full_rgb_range_output && rsx::fcmp(avconfig.gamma, 1.f) && avconfig.stereo_mode == stereo_render_mode_options::disabled)
```

File: rpcs3/Emu/RSX/GL/GLPresent.cpp [hunk 17]
```diff
@@ -181,7 +181,7 @@ void GLGSRender::flip(const rsx::display_flip_info_t& info)
 		const auto image_to_flip_ = get_present_source(&present_info, avconfig);
 		image_to_flip = image_to_flip_->id();
 
-		if (avconfig._3d) [[unlikely]]
+		if (avconfig.stereo_mode == stereo_render_mode_options::disabled) [[unlikely]]
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLPresent.cpp:184]: I think this is a typo
```suggestion
		if (avconfig.stereo_mode != stereo_render_mode_options::disabled) [[unlikely]]
```

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 18]
```diff
@@ -467,30 +467,69 @@ namespace gl
...
-			"	if (stereo_image_count == 2)\n"
+			"	if (stereo_image_count == 1)\n"
+			"	{\n"
+			"		if(stereo_display_mode == 1)\n" // anaglyph
+			"		{\n"
+			"			left = texture(fs0, tc0 * vec2(1.f, 0.4898f));\n"
+			"			right = texture(fs0, (tc0 * vec2(1.f, 0.4898f)) + vec2(0.f, 0.510204f));\n"
+			"			return vec4(left.r, right.g, right.b, 1.f);\n"
+			"		}\n"
+			"		if(stereo_display_mode == 2)\n" // side-by-side
```
Comment (kd-11) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:493]: Declare these as defines just before the matrices. We'll need the following enums:
```C
#define STEREO_MODE_DISABLED 0
#define STEREO_MODE_ANAGLYPH 1
#define STEREO_MODE_SIDE_BY_SIDE 2
#define STEREO_MODE_OVER_UNDER 3
```

Also, replace the if-else chain with a switch-case block. It will be much easier to read.

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 19]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"	}\n"
+			"	else if (stereo_image_count == 2)\n"
+			"	{\n"
+			"		switch (stereo_display_mode)\n"
+			"		{\n"
+			"		case STEREO_MODE_ANAGLYPH:\n"
+			"			left = texture(fs0, tc0);\n"
+			"			right = texture(fs1, tc0);\n"
+			"			return vec4(left.r, right.g, right.b, 1.f);\n"
+			"		  break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:520]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 20]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"			if (tc0.x < 0.5) return texture(fs0,(tc0 * sbs_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * sbs_multi_matrix) + vec2(-1.f,0.f));\n"
+			"		  break;\n"
+			"		case STEREO_MODE_OVER_UNDER:\n"
+			"			if (tc0.y < 0.5) return texture(fs0,(tc0 * ou_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * ou_multi_matrix) + vec2(0.f,-1.f));\n"
+			"		  break;\n"
+			"		default:\n" // undefined behavior
+			"		  return texture(fs0,tc0);\n"
+			"		  break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:531]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 21]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"		case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"			if (tc0.x < 0.5) return texture(fs0,(tc0 * sbs_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * sbs_multi_matrix) + vec2(-1.f,0.f));\n"
+			"		  break;\n"
+			"		case STEREO_MODE_OVER_UNDER:\n"
+			"			if (tc0.y < 0.5) return texture(fs0,(tc0 * ou_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * ou_multi_matrix) + vec2(0.f,-1.f));\n"
+			"		  break;\n"
+			"		default:\n" // undefined behavior
+			"		  return texture(fs0,tc0);\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:530]: indentation

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 22]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"		{\n"
+			"		case STEREO_MODE_ANAGLYPH:\n"
+			"			left = texture(fs0, tc0);\n"
+			"			right = texture(fs1, tc0);\n"
+			"			return vec4(left.r, right.g, right.b, 1.f);\n"
+			"		  break;\n"
+			"		case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"			if (tc0.x < 0.5) return texture(fs0,(tc0 * sbs_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * sbs_multi_matrix) + vec2(-1.f,0.f));\n"
+			"		  break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:524]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 23]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"				if (tc0.x < 0.5) return texture(fs0, tc0* sbs_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* sbs_single_matrix) + vec2(-1.f, 0.510204f));\n"
+			"				break;"
+			"			case STEREO_MODE_OVER_UNDER:\n"
+			"				if (tc0.y < 0.5) return texture(fs0, tc0* ou_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* ou_single_matrix) + vec2(0.f, 0.020408f) );\n"
+			"				break;"
+			"			default:\n" // undefined behavior
+			"				return texture(fs0,tc0);\n"
+			"				break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:509]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 24]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"				return vec4(left.r, right.g, right.b, 1.f);\n"
+			"				break;"
+			"			case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"				if (tc0.x < 0.5) return texture(fs0, tc0* sbs_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* sbs_single_matrix) + vec2(-1.f, 0.510204f));\n"
+			"				break;"
+			"			case STEREO_MODE_OVER_UNDER:\n"
+			"				if (tc0.y < 0.5) return texture(fs0, tc0* ou_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* ou_single_matrix) + vec2(0.f, 0.020408f) );\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:506]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 25]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
+			"		{\n"
+			"			case STEREO_MODE_ANAGLYPH:\n"
+			"				left = texture(fs0, tc0 * vec2(1.f, 0.4898f));\n"
+			"				right = texture(fs0, (tc0 * vec2(1.f, 0.4898f)) + vec2(0.f, 0.510204f));\n"
+			"				return vec4(left.r, right.g, right.b, 1.f);\n"
+			"				break;"
+			"			case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"				if (tc0.x < 0.5) return texture(fs0, tc0* sbs_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* sbs_single_matrix) + vec2(-1.f, 0.510204f));\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:502]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 26]
```diff
@@ -467,30 +467,77 @@ namespace gl
...
 			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
+			"		switch (stereo_display_mode)\n"
+			"		{\n"
+			"			case STEREO_MODE_ANAGLYPH:\n"
+			"				left = texture(fs0, tc0 * vec2(1.f, 0.4898f));\n"
+			"				right = texture(fs0, (tc0 * vec2(1.f, 0.4898f)) + vec2(0.f, 0.510204f));\n"
+			"				return vec4(left.r, right.g, right.b, 1.f);\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:498]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 27]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
 			"	{\n"
-			"		left = texture(fs0, tc0);\n"
-			"		right = texture(fs1, tc0);\n"
+			"		switch (stereo_display_mode)\n"
+			"		{\n"
+			"			case STEREO_MODE_ANAGLYPH:\n"
+			"				left = texture(fs0, tc0 * vec2(1.f, 0.4898f));\n"
+			"				right = texture(fs0, (tc0 * vec2(1.f, 0.4898f)) + vec2(0.f, 0.510204f));\n"
+			"				return vec4(left.r, right.g, right.b, 1.f);\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:927]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 28]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"			case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"				if (tc0.x < 0.5) return texture(fs0, tc0* sbs_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* sbs_single_matrix) + vec2(-1.f, 0.510204f));\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:931]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 29]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"			case STEREO_MODE_OVER_UNDER:\n"
+			"				if (tc0.y < 0.5) return texture(fs0, tc0* ou_single_matrix);\n"
+			"				else             return texture(fs0, (tc0* ou_single_matrix) + vec2(0.f, 0.020408f) );\n"
+			"				break;"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:935]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 30]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"		  default:\n" // undefined behavior
+			"				return texture(fs0,tc0);\n"
+			"				break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:938]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 31]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"	}\n"
+			"	else if (stereo_image_count == 2)\n"
+			"	{\n"
+			"		switch (stereo_display_mode)\n"
+			"		{\n"
+			"		case STEREO_MODE_ANAGLYPH:\n"
+			"			left = texture(fs0, tc0);\n"
+			"			right = texture(fs1, tc0);\n"
+			"			return vec4(left.r, right.g, right.b, 1.f);\n"
+			"			break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:949]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 32]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"		case STEREO_MODE_SIDE_BY_SIDE:\n"
+			"			if (tc0.x < 0.5) return texture(fs0,(tc0 * sbs_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * sbs_multi_matrix) + vec2(-1.f,0.f));\n"
+			"			break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:953]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 33]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"		case STEREO_MODE_OVER_UNDER:\n"
+			"			if (tc0.y < 0.5) return texture(fs0,(tc0 * ou_multi_matrix));\n"
+			"			else             return texture(fs1,(tc0 * ou_multi_matrix) + vec2(0.f,-1.f));\n"
+			"			break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:957]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 34]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"		default:\n" // undefined behavior
+			"			return texture(fs0,tc0);\n"
+			"			break;\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:960]: obsolete. this line should be removed

File: rpcs3/Emu/RSX/VK/VKOverlays.cpp [hunk 35]
```diff
@@ -893,33 +893,81 @@ namespace vk
...
+			"		}\n"
 			"	}\n"
 			"	else\n"
 			"	{\n"
 			"		vec2 coord_left = tc0 * vec2(1.f, 0.4898f);\n"
 			"		vec2 coord_right = coord_left + vec2(0.f, 0.510204f);\n"
 			"		left = texture(fs0, coord_left);\n"
 			"		right = texture(fs0, coord_right);\n"
+			" 	return vec4(left.r, right.g, right.b, 1.);\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKOverlays.cpp:969]: indentation

**Review by kd-11 — APPROVED** [review]
There are still some minor formatting and code structure problems, but nothing serious. I may follow up with a refactor to clean that up soon after merge.
